[Wasm-GC] Implement array.new_canon_data

catamorphism opened the pull request:
#### 346a6a54590426e49ac55ab27e6bb55228bc23d0
<pre>
[Wasm-GC] Implement array.new_canon_data
<a href="https://bugs.webkit.org/show_bug.cgi?id=251329">https://bugs.webkit.org/show_bug.cgi?id=251329</a>

Reviewed by Justin Michaud.

Implement the `array.new_canon_data` instruction, which creates a
new array initialized from a data segment (as specified in the GC
proposal MVP).

* JSTests/wasm/gc/array_new_data.js: Added.
(module):
(check):
(assertFails):
(testArrayNewCanonData):
(testBadDataSegment):
(testBadOffset.check):
(testBadOffset):
(testReadOutOfBounds):
(testInt32Overflow):
(testTypeErrors):
(testBadOperands):
* JSTests/wasm/wasm.json:
* Source/JavaScriptCore/wasm/WasmAirIRGeneratorBase.h:
(JSC::Wasm::ExpressionType&gt;::addArrayNew):
(JSC::Wasm::ExpressionType&gt;::addArrayNewData):
(JSC::Wasm::ExpressionType&gt;::addArrayGet):
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::B3IRGenerator::addArrayNewData):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
* Source/JavaScriptCore/wasm/WasmCallingConvention.h:
(JSC::Wasm::WasmCallingConvention::numberOfStackResults const):
* Source/JavaScriptCore/wasm/WasmExceptionType.h:
(JSC::Wasm::isTypeErrorExceptionType):
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser&lt;Context&gt;::parseExpression):
* Source/JavaScriptCore/wasm/WasmInstance.cpp:
(JSC::Wasm::Instance::copyDataSegment):
* Source/JavaScriptCore/wasm/WasmInstance.h:
* Source/JavaScriptCore/wasm/WasmLLIntBuiltin.h:
* Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp:
(JSC::Wasm::LLIntGenerator::callInformationForCaller):
(JSC::Wasm::LLIntGenerator::addArrayNewData):
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/wasm/WasmOperations.h:
* Source/JavaScriptCore/wasm/WasmOperationsInlines.h:
(JSC::Wasm::createArrayFromDataSegment):
(JSC::Wasm::arrayNewData):
* Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
(JSC::LLInt::WASM_SLOW_PATH_DECL):
* Source/JavaScriptCore/wasm/WasmTypeDefinition.cpp:
(JSC::Wasm::TypeInformation::signatureForLLIntBuiltin):
(JSC::Wasm::TypeInformation::TypeInformation):
* Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
* Source/JavaScriptCore/wasm/wasm.json:

Canonical link: <a href="https://commits.webkit.org/261983@main">https://commits.webkit.org/261983@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/926b8628b73b09914c5f148b18b938ed472aeb4a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/113197 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/22330 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/1870 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/4973 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/121657 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/23700 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/13508 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/6206 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/118984 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/17606 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100864 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/106284 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/99560 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/1404 "4 failures") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/46647 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/101431 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/14629 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/1445 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/98902 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/12791 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/15341 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/10776 "10 flakes 4 failures") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/102992 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/20633 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/53436 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/32113 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/43 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/17191 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/111044 "Built successfully") | 
| | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/27413 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/bytecode/BytecodeList.rb:
```diff
@@ -1878,6 +1878,15 @@
+        offset: VirtualRegister,
+        typeIndex: unsigned,
+        dataIndex: unsigned,
+    }
```
We should implement these things as a builtin call instead of adding a new byte code. See call_builtin implementation.

Constellation commented on Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
```diff
@@ -415,6 +415,22 @@ WASM_SLOW_PATH_DECL(array_new)
+    uint32_t arraySize = READ(instruction.m_size).unboxedUInt32();
+    uint32_t offset = READ(instruction.m_offset).unboxedUInt32();
+
+    auto result = Wasm::operationWasmArrayNewData(instance,
```
This is not OK. We should not call operation function from C++. You need to extract code into WasmOperationsInline.h

Constellation reviewed (changes requested):
Added some comments.

Constellation commented on Source/JavaScriptCore/wasm/WasmOperations.cpp:
```diff
@@ -932,6 +932,11 @@ JSC_DEFINE_JIT_OPERATION(operationWasmArrayNew, EncodedJSValue, (Instance* insta
 }
 
+JSC_DEFINE_JIT_OPERATION(operationWasmArrayNewData, EncodedJSValue, (Instance* instance, uint32_t typeIndex, uint32_t dataSegmentIndex, uint32_t arraySize, uint32_t offset))
+{
```
You need to put 

```
    CallFrame* callFrame = DECLARE_WASM_CALL_FRAME(instance);
    VM& vm = instance->vm();
    NativeCallFrameTracer tracer(vm, callFrame);
```

If the following operations can cause GC / error etc. And in this case, they can happen.

Constellation commented on Source/JavaScriptCore/wasm/WasmOperationsInlines.h:
```diff
@@ -98,6 +98,75 @@ inline EncodedJSValue arrayNew(Instance* instance, uint32_t typeIndex, uint32_t
+
+    // Copy the data from the segment into the temp `values` vector
+    if (!instance->copyDataSegment(dataSegmentIndex, offset, arrayLengthInBytes, tempValues)) {
+        throwTypeError(globalObject, scope, "Offset plus array size is out of bounds for data segment"_s);
```
Is it possible to add a test for this case?

catamorphism replied:
I realized that this case actually shouldn't be there, because if this code is being called from B3 or Air, the bounds check was already done and the exception was thrown if necessary. I reorganized the code so that this just returns false without throwing the error, and added a check to the `call_builtin` case for `ArrayNewData` instead, so that the caller throws the exception whether the operation gets called from B3, Air, or LLInt. Let me know what you think.

Constellation commented on Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
```diff
@@ -2726,6 +2727,42 @@ auto B3IRGenerator::addArrayNewDefault(uint32_t typeIndex, ExpressionType size,
+
+    // FIXME: Emit this inline.
+    // https://bugs.webkit.org/show_bug.cgi?id=245405
+    result = push(m_currentBlock->appendNew<CCallValue>(m_proc, toB3Type(Types::Arrayref), origin(),
```
You need to use `callWasmOperation`.

Constellation commented on Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
```diff
@@ -2726,6 +2727,42 @@ auto B3IRGenerator::addArrayNewDefault(uint32_t typeIndex, ExpressionType size,
+        });
+    }
+
+    // FIXME: Emit this inline.
```
I don't think this function should be inlined since it is very complex, correct?

catamorphism replied:
Since it ultimately calls `memcpy()`, I suppose you're right.

Constellation commented on Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
```diff
@@ -579,6 +579,7 @@ class B3IRGenerator {
     PartialResult WARN_UNUSED_RETURN addArrayNew(uint32_t index, ExpressionType size, ExpressionType value, ExpressionType& result);
     PartialResult WARN_UNUSED_RETURN addArrayNewDefault(uint32_t index, ExpressionType size, ExpressionType& result);
     PartialResult WARN_UNUSED_RETURN addArrayGet(ExtGCOpType arrayGetKind, uint32_t typeIndex, ExpressionType arrayref, ExpressionType index, ExpressionType& result);
+    PartialResult WARN_UNUSED_RETURN addArrayNewData(uint32_t typeIndex, uint32_t dataIndex, ExpressionType size, ExpressionType offset, ExpressionType& result);
```
Please add stub to WasmBBQJIT.cpp too.

Constellation commented on Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
```diff
@@ -2726,6 +2727,26 @@ auto B3IRGenerator::addArrayNewDefault(uint32_t typeIndex, ExpressionType size,
+        m_currentBlock->appendNew<Const32Value>(m_proc, origin(), dataIndex),
+        get(arraySize), get(offset));
+
+    result = push(resultValue);
```
Why pushing it before checking?

justinmichaud reviewed (approved):
r=me

takikawa commented on Source/JavaScriptCore/wasm/WasmCallingConvention.h:
```diff
@@ -162,6 +162,7 @@ class WasmCallingConvention {
             case TypeKind::Funcref:
             case TypeKind::RefNull:
             case TypeKind::Ref:
+            case TypeKind::Arrayref:
```
I think that this isn't quite the right change here, because `arrayref` as a heap type should typically not appear as an actual type kind (it should be a `RefNull`/`Ref`). The reason it's needed is I think due to a use of `Types::Arrayref` below which can be changed.

takikawa commented on Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp:
```diff
@@ -754,7 +756,6 @@ auto LLIntGenerator::callInformationForCaller(const FunctionSignature& signature
         case TypeKind::Struct:
         case TypeKind::Structref:
         case TypeKind::Array:
-        case TypeKind::Arrayref:
```
Ditto previous comment.

takikawa commented on Source/JavaScriptCore/wasm/WasmTypeDefinition.cpp:
```diff
@@ -853,6 +855,7 @@ TypeInformation::TypeInformation()
     // FIXME: The Ref type here is only used to compute call information, a heap type of "any" would be better once that's added.
     m_I32_RefI32I32 = m_typeSet.template add<FunctionParameterTypes>(FunctionParameterTypes { { Wasm::Types::I32 }, { Wasm::Type { Wasm::TypeKind::Ref, static_cast<TypeIndex>(Wasm::TypeKind::Externref) }, Wasm::Types::I32, Wasm::Types::I32 } }).iterator->key;
     m_Ref_RefI32I32 = m_typeSet.template add<FunctionParameterTypes>(FunctionParameterTypes { { Wasm::Type { Wasm::TypeKind::Ref, static_cast<TypeIndex>(Wasm::TypeKind::Externref) } },  { Wasm::Type { Wasm::TypeKind::Ref, static_cast<TypeIndex>(Wasm::TypeKind::Externref) }, Wasm::Types::I32, Wasm::Types::I32 } }).iterator->key;
+    m_Arrayref_I32I32I32I32 = m_typeSet.template add<FunctionParameterTypes>(FunctionParameterTypes { { Wasm::Types::Arrayref }, { Wasm::Types::I32, Wasm::Types::I32, Wasm::Types::I32, Wasm::Types::I32 } }).iterator->key;
```
I think `Wasm::Types::Arrayref` should be `Wasm::Type { Wasm::TypeKind::Ref, static_cast<TypeIndex>(Wasm::TypeKind::Arrayref) }` (like some of the other cases above).

Alternatively you could define a helper for this like `arrayrefType()` in `WasmFormat.h` (see `eqrefType()` as an example).

catamorphism committed:
[Wasm-GC] Implement array.new_canon_data
https://bugs.webkit.org/show_bug.cgi?id=251329

Reviewed by Justin Michaud.

Implement the `array.new_canon_data` instruction, which creates a
new array initialized from a data segment (as specified in the GC
proposal MVP).

* JSTests/wasm/gc/array_new_data.js: Added.
(module):
(check):
(assertFails):
(testArrayNewCanonData):
(testBadDataSegment):
(testBadOffset.check):
(testBadOffset):
(testReadOutOfBounds):
(testInt32Overflow):
(testTypeErrors):
(testBadOperands):
* JSTests/wasm/wasm.json:
* Source/JavaScriptCore/wasm/WasmAirIRGeneratorBase.h:
(JSC::Wasm::ExpressionType>::addArrayNew):
(JSC::Wasm::ExpressionType>::addArrayNewData):
(JSC::Wasm::ExpressionType>::addArrayGet):
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::B3IRGenerator::addArrayNewData):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
* Source/JavaScriptCore/wasm/WasmCallingConvention.h:
(JSC::Wasm::WasmCallingConvention::numberOfStackResults const):
* Source/JavaScriptCore/wasm/WasmExceptionType.h:
(JSC::Wasm::isTypeErrorExceptionType):
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser<Context>::parseExpression):
* Source/JavaScriptCore/wasm/WasmInstance.cpp:
(JSC::Wasm::Instance::copyDataSegment):
* Source/JavaScriptCore/wasm/WasmInstance.h:
* Source/JavaScriptCore/wasm/WasmLLIntBuiltin.h:
* Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp:
(JSC::Wasm::LLIntGenerator::callInformationForCaller):
(JSC::Wasm::LLIntGenerator::addArrayNewData):
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/wasm/WasmOperations.h:
* Source/JavaScriptCore/wasm/WasmOperationsInlines.h:
(JSC::Wasm::createArrayFromDataSegment):
(JSC::Wasm::arrayNewData):
* Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
(JSC::LLInt::WASM_SLOW_PATH_DECL):
* Source/JavaScriptCore/wasm/WasmTypeDefinition.cpp:
(JSC::Wasm::TypeInformation::signatureForLLIntBuiltin):
(JSC::Wasm::TypeInformation::TypeInformation):
* Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
* Source/JavaScriptCore/wasm/wasm.json:

Canonical link: https://commits.webkit.org/261983@main
